torch/csrc/jit/tensorexpr/ir_cloner.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+#include <c10/core/ScalarType.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <vector>
+
+#include <torch/csrc/jit/tensorexpr/ir_mutator.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+class Add;
+class Sub;
+class Mul;
+class Div;
+class Mod;
+class Max;
+class Min;
+class And;
+class Or;
+class Xor;
+class Lshift;
+class Rshift;
+class CompareSelect;
+
+#define IMM_DECLARE(Type, Name) class Name##Imm;
+AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, IMM_DECLARE);
+#undef IMM_DECLARE
+
+class Cast;
+class BitCast;
+class Var;
+class Buf;
+class Ramp;
+class Load;
+class For;
+class Block;
+class Store;
+class Broadcast;
+class IfThenElse;
+class ExprHandle;
+class Expr;
+class Intrinsics;
+class Allocate;
+class Free;
+class Let;
+class Cond;
+class Stmt;
+class Term;
+class Polynomial;
+class RoundOff;
+class MaxTerm;
+class MinTerm;
+class ReduceOp;
+class AtomicAdd;
+class SyncThreads;
+class ExternalCall;
+
+class TORCH_API IRCloner : public IRMutator {
```

Comment:
Should we separate IRCloner from IRMutator? They server for very different purposes and it looks weird that we do `clone` in the name of `mutate`. Since the IRCloner has its own mutations on expressions and stmts now, I guess there's not much code it can reuse from IRMutator?

Reply:
Actually, I considered this and I personally didn't have a strong preference either way. And it was just simpler to keep it same as before.

If you have a strong preference to doing this, I can do it in a separate PR.